File: Day2/day2.dart
import "dart:core";
import "dart:io";

void main() async {
    String contents = await File("./day2_input.txt").readAsString();
    List<int> numbers = [];
    var input = contents.split("\n");

    print("Part one: ${partOne(input)}");
    print("Part two: ${partTwo(input)}");
}

int partOne(List<String> input) {
    int forward = 0;
    int depth = 0;

    for(String inst in input) {
        List<String> parts = inst.split(" ");
        var value = int.parse(parts[1]);
        if (parts[0] == "forward") {
            forward += value; 
        } else if (parts[0] == "up") {
            depth -= value;
        } else if (parts[0] == "down") {
            depth += value;
        }
    }

    return forward * depth;
}

int partTwo(List<String> input) {
    int forward = 0;
    int depth = 0;
    int aim = 0;

    for(String inst in input) {
        List<String> parts = inst.split(" ");
        var value = int.parse(parts[1]);
        if (parts[0] == "forward") {
            forward += value; 
            depth += aim * value;
        } else if (parts[0] == "up") {
            aim -= value;
        } else if (parts[0] == "down") {
            aim += value;
        }
    }

    return forward * depth;
}

File: Day2/day2.ts
export function partOne(input: string[]): number {
    let depth = 0;
    let forward = 0;
    input.forEach(inst => {
        let parts = inst.split(' ');
        switch (parts[0]) {
            case 'forward':
                forward += parseInt(parts[1]);
                break;
            case 'up':
                depth -= parseInt(parts[1]);
                break;
            case 'down':
                depth += parseInt(parts[1]);
                break;
        }
    });

    return depth * forward;
}

export function partTwo(input: string[]): number {
    let aim = 0;
    let forward = 0;
    let depth = 0;
    input.forEach(inst => {
        let parts = inst.split(' ');
        switch (parts[0]) {
            case 'forward':
                let num = parseInt(parts[1]);
                forward += num;
                depth += aim * num;
                break;
            case 'up':
                aim -= parseInt(parts[1]);
                break;
            case 'down':
                aim += parseInt(parts[1]);
                break;
        }
    });

    return depth * forward;
}